components/sidebar/sidebar_service.cc

```diff
@@ -225,75 +354,134 @@ SidebarService::ShowSidebarOption SidebarService::GetSidebarShowOption() const {
+            continue;
+          }
+          // Valid built-in item, add it
+          items_.emplace_back(*std::make_move_iterator(iter));
+          default_items_to_add.erase(iter);
+          continue;
+        }
+      }
+      // Deserialize custom item
+      std::string url;
+      if (const auto* value = item.FindStringKey(kSidebarItemURLKey)) {
+        url = *value;
+      } else {
+        continue;
+      }
+      // Open in panel for custom items is not yet supported
+      bool open_in_panel = false;
+      std::string title;
+      if (const auto* value = item.FindStringKey(kSidebarItemTitleKey)) {
+        title = *value;
+      }
+      items_.push_back(SidebarItem::Create(
+          GURL(url), base::UTF8ToUTF16(title), type,
+          SidebarItem::BuiltInItemType::kNone, open_in_panel));
     }
+  }
 
-    // Title can be updated later.
-    std::string title;
-    if (const auto* value = item.FindStringKey(kSidebarItemTitleKey)) {
-      title = *value;
+  //
+  // Add built-in items which haven't been shown or hidden.
+  //
+  // Don't consider built-in items that the user has already hidden.
+  auto* hidden_built_in_preference =
+      prefs_->FindPreference(kSidebarHiddenBuiltInItems);
+  if (!hidden_built_in_preference->IsDefaultValue()) {
+    for (const auto& item : hidden_built_in_preference->GetValue()->GetList()) {
+      // Don't show this built-in item
+      const auto id = static_cast<SidebarItem::BuiltInItemType>(item.GetInt());
+      DVLOG(2) << "hide built-in item with id: " << item.GetInt();
+      auto iter =
+          std::find_if(default_items_to_add.begin(), default_items_to_add.end(),
+                       [id](const auto& default_item) {
+                         return default_item.built_in_item_type ==
+                                static_cast<SidebarItem::BuiltInItemType>(id);
+                       });
+      if (iter != default_items_to_add.end()) {
+        default_items_to_add.erase(iter);
+      } else {
+        DLOG(ERROR) << "Asked to hide an item that was already asked to show. "
+                       "This indicates something is wrong with the "
+                       "serialization process. Id was: "
+                    << item.GetInt();
+      }
     }
-
-    DCHECK(type != SidebarItem::Type::kTypeBuiltIn);
-    items_.push_back(SidebarItem::Create(
-        GURL(url), base::UTF8ToUTF16(title), type,
-        SidebarItem::BuiltInItemType::kNone, open_in_panel));
   }
-}
 
-// For now, only builtin history item is blocked.
-bool SidebarService::IsBlockedBuiltInItem(const SidebarItem& item) const {
-  if (!IsBuiltInType(item))
-    return false;
-  return item.built_in_item_type == SidebarItem::BuiltInItemType::kHistory;
+  // Add the items the user has never seen (or never persisted).
+  // Get the initial order of items so that we can attempt to
+  // insert at the intended order.
+  auto default_item_types = GetDefaultBuiltInItemTypes();
```

Comment:
If GetDefaultBuiltInItemTypes returns const&, const auto& will be better.
```suggestion
  const auto& default_item_types = GetDefaultBuiltInItemTypes();
```